src/librustc_middle/ty/sty.rs

```diff
@@ -2371,27 +2398,25 @@ impl<'tcx> Const<'tcx> {
                 // and we use the original type, so nothing from `substs`
                 // (which may be identity substs, see above),
                 // can leak through `val` into the const we return.
-                Ok(val) => Const::from_value(tcx, val, self.ty),
-                Err(ErrorHandled::TooGeneric | ErrorHandled::Linted) => self,
-                Err(ErrorHandled::Reported(ErrorReported)) => tcx.const_error(self.ty),
+                Ok(val) => Some(Ok(val)),
+                Err(ErrorHandled::TooGeneric | ErrorHandled::Linted) => None,
+                Err(ErrorHandled::Reported(e)) => Some(Err(e)),
             }
         } else {
-            self
+            None
         }
     }
+}
 
+impl<'tcx> Const<'tcx> {
     #[inline]
     pub fn try_eval_bool(&self, tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>) -> Option<bool> {
-        self.try_eval_bits(tcx, param_env, tcx.types.bool).and_then(|v| match v {
-            0 => Some(false),
-            1 => Some(true),
-            _ => None,
-        })
+        self.val.eval(tcx, param_env).try_to_bool()
```

Comment:
this reduces interning

Reply:
This also reduces `layout_of` calls, as we use `try_to_bool` instead of passing `tcx.types.bool` and fetching the layout for that.

Reply:
Oh, `val` has type `ConstKind`, not `ConstValue`... of course.
Man the names are really confusing.^^

Reply:
:laughing: maybe I should clean up more before I do the `mir::Const` thing (this PR is already part of that cleanup process)

Reply:
Yes please. :D